utils/bincompat/generate.py

```diff
@@ -0,0 +1,308 @@
+#!/usr/bin/env python
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+
+"""Generate run scripts for Unikraft bincompat applications.
+
+Use QEMU, Firecracker and Krafkit for running.
+"""
+
+import sys
+import os
+import stat
+import yaml
+
+
+TEMPLATE_RUN_QEMU_HEADER = """#!/bin/sh
+
+kernel="{}"
+cmd="{}"
+
+if test $# -eq 1; then
+    kernel="$1"
+fi
+"""
+
+RUN_COMMON_NET_COMMANDS = """
+# Remove previously created network interfaces.
+sudo ip link set dev tap0 down 2> /dev/null
+sudo ip link del dev tap0 2> /dev/null
+sudo ip link set dev virbr0 down 2> /dev/null
+sudo ip link del dev virbr0 2> /dev/null
+"""
+
+RUN_QEMU_NET_COMMANDS = """
+# Create bridge interface for QEMU networking.
+sudo ip link add dev virbr0 type bridge
+sudo ip address add 172.44.0.1/24 dev virbr0
+sudo ip link set dev virbr0 up
+"""
+
+RUN_FIRECRACKER_NET_COMMANDS = """
+# Create tap interface for Firecracker networking.
+sudo ip tuntap add dev tap0 mode tap
+sudo ip address add 172.44.0.1/24 dev tap0
+sudo ip link set dev tap0 up
+"""
+
+RUN_KRAFT_NET_COMMANDS = """
+# Create bridge interface for KraftKit networking.
+sudo kraft net create -n 172.44.0.1/24 virbr0
+"""
+
+TEMPLATE_RUN_FIRECRACKER_HEADER = """#!/bin/sh
+
+config="{}"
+
+if test $# -eq 1; then
+    config="$1"
+fi
+"""
+
+RUN_FIRECRACKER_PRE_TRAILER = """
+# Remove previously created files.
+sudo rm -f /tmp/firecracker.log
+> /tmp/firecracker.log
+sudo rm -f /tmp/firecracker.socket
+"""
+
+RUN_FIRECRACKER_COMMAND = """firecracker-x86_64 \\
+        --api-sock /tmp/firecracker.socket \\
+        --config-file "$config"
+"""
+
+RUN_KRAFT_HEADER = """#!/bin/sh
+"""
+
+RUN_KILL_COMMANDS = """
+# Clean up any previous instances.
+sudo pkill -f qemu-system 2> /dev/null
+sudo pkill -f firecracker 2> /dev/null
+sudo kraft stop --all 2> /dev/null
+sudo kraft rm --all 2> /dev/null
+"""
+
+CONFIG = "config.yaml"
+KRAFTCONFIG = "Kraftfile"
+KERNELS = "../../kernels"
+RUNDIR = "."
```

Comment:
Can we somehow use absolute paths for this?
It might fail for some more specific file layout (e.g. #33, where the workdir will be `library/nginx/version/`, and the path should be `../../../kernels`).

Reply:
Yes, you're right. Lemme think about that.